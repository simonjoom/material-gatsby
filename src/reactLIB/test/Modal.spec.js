import React from 'react';
import { shallow, mount } from 'enzyme';
import Modal from '../src/Modal';
import mocker from './helper/mocker';

describe('<Modal />', () => {
  let wrapper;
  const modalMock = jest.fn();
  const restore = mocker('modal', modalMock);

  afterAll(() => {
    restore();
  });

  const trigger = <button>click</button>;
  const children = (
    <div>
      <h1>Hello</h1>
    </div>
  );
  const header = 'Modal header';
  const modalOptions = {
    dismissible: true,
    opacity: 0.4
  };

  let renderedWrapper;

  describe('renders a modal', () => {
    beforeEach(() => {
      wrapper = mount(
        <Modal trigger={trigger} modalOptions={modalOptions} header={header}>
          {children}
        </Modal>
      );

      renderedWrapper = document.body.lastElementChild.innerHTML;
    });

    afterEach(() => {
      modalMock.mockClear();
      document.body.removeChild(document.body.lastElementChild);
    });

    test('has children', () => {
      expect(renderedWrapper).toContain(header);
      expect(wrapper).toMatchSnapshot();
    });

    test('has a footer', () => {
      expect(renderedWrapper).toContain('modal-footer');
    });
  });

  describe('without a trigger', () => {
    beforeEach(() => {
      wrapper = mount(<Modal modalOptions={modalOptions}>{children}</Modal>);
    });

    afterEach(() => {
      modalMock.mockClear();
      document.body.removeChild(document.body.lastElementChild);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find(Modal).length).toEqual(1);
    });
  });

  describe('controlled modal with `open` prop', () => {
    let testModal;
    beforeEach(() => {
      testModal = shallow(
        <Modal modalOptions={{ one: 1 }} open>
          {children}
        </Modal>
      );
    });

    afterEach(() => {
      modalMock.mockClear();
      document.body.removeChild(document.body.lastElementChild);
    });

    test('mounts opened', () => {
      // once in mount and twice in #showModal
      expect(modalMock).toHaveBeenCalledTimes(3);
      expect(wrapper).toMatchSnapshot();
    });

    test('open on prop change', () => {
      testModal.setProps({ open: true });
      expect(modalMock).toHaveBeenCalledTimes(3);
      // no trigger is defined, modal should be configured in constructor
      expect(modalMock.mock.calls[0]).toEqual([{ one: 1 }]);
      // showModal initializes the modal again
      expect(modalMock.mock.calls[1]).toEqual([{ one: 1 }]);
      expect(modalMock).toHaveBeenLastCalledWith('open');
    });

    test('closes on prop change', () => {
      testModal.setProps({ open: false });
      expect(modalMock).toHaveBeenCalledTimes(4);
      // no trigger is defined, modal should be configured in constructor
      expect(modalMock.mock.calls[0]).toEqual([{ one: 1 }]);
      // open prop is set, so showModal is called
      expect(modalMock.mock.calls[1]).toEqual([{ one: 1 }]);
      expect(modalMock.mock.calls[2]).toEqual(['open']);
      expect(modalMock).toHaveBeenLastCalledWith('close');
    });
  });

  describe('renders a trigger', () => {
    beforeEach(() => {
      wrapper = shallow(
        <Modal trigger={trigger} modalOptions={modalOptions} header={header}>
          {children}
        </Modal>
      );
    });

    afterEach(() => {
      modalMock.mockClear();
      document.body.removeChild(document.body.lastElementChild);
    });

    test('renders', () => {
      expect(wrapper.find('button').length).toEqual(1);
      expect(wrapper).toMatchSnapshot();
    });

    test('initializes with modalOptions', () => {
      wrapper.find('button').simulate('click');
      expect(modalMock).toHaveBeenCalledWith(modalOptions);
    });
  });
});
