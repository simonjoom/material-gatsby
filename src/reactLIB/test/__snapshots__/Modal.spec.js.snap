// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> controlled modal with \`open\` prop mounts opened 1`] = `
<Modal
  actions={
    Array [
      <Button
        flat={true}
        modal="close"
        node="button"
        waves="light"
      >
        Close
      </Button>,
    ]
  }
  bottomSheet={false}
  fixedFooter={false}
  modalOptions={
    Object {
      "dismissible": true,
      "opacity": 0.4,
    }
  }
>
  <div>
    <div
      className="modal"
      id="modal_2"
    >
      <div
        className="modal-content"
      >
        <h4 />
        <div>
          <h1>
            Hello
          </h1>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <Button
          flat={true}
          key=".0"
          modal="close"
          node="button"
          waves="light"
        >
          <button
            className="btn waves-effect waves-light btn-flat modal-action modal-close"
            disabled={false}
          >
            Close
          </button>
        </Button>
      </div>
    </div>
  </div>
</Modal>
`;

exports[`<Modal /> renders a modal has children 1`] = `
<Modal
  actions={
    Array [
      <Button
        flat={true}
        modal="close"
        node="button"
        waves="light"
      >
        Close
      </Button>,
    ]
  }
  bottomSheet={false}
  fixedFooter={false}
  header="Modal header"
  modalOptions={
    Object {
      "dismissible": true,
      "opacity": 0.4,
    }
  }
  trigger={
    <button>
      click
    </button>
  }
>
  <div>
    <button
      onClick={[Function]}
    >
      click
    </button>
    <div
      className="modal"
      id="modal_0"
    >
      <div
        className="modal-content"
      >
        <h4>
          Modal header
        </h4>
        <div>
          <h1>
            Hello
          </h1>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <Button
          flat={true}
          key=".0"
          modal="close"
          node="button"
          waves="light"
        >
          <button
            className="btn waves-effect waves-light btn-flat modal-action modal-close"
            disabled={false}
          >
            Close
          </button>
        </Button>
      </div>
    </div>
  </div>
</Modal>
`;

exports[`<Modal /> renders a trigger renders 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    click
  </button>
  <Portal />
</div>
`;

exports[`<Modal /> without a trigger renders 1`] = `
<Modal
  actions={
    Array [
      <Button
        flat={true}
        modal="close"
        node="button"
        waves="light"
      >
        Close
      </Button>,
    ]
  }
  bottomSheet={false}
  fixedFooter={false}
  modalOptions={
    Object {
      "dismissible": true,
      "opacity": 0.4,
    }
  }
>
  <div>
    <div
      className="modal"
      id="modal_2"
    >
      <div
        className="modal-content"
      >
        <h4 />
        <div>
          <h1>
            Hello
          </h1>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <Button
          flat={true}
          key=".0"
          modal="close"
          node="button"
          waves="light"
        >
          <button
            className="btn waves-effect waves-light btn-flat modal-action modal-close"
            disabled={false}
          >
            Close
          </button>
        </Button>
      </div>
    </div>
  </div>
</Modal>
`;
